File: business-logic/classic-scanlog-core/tests/semantic_conformance/message_logging.rs
//! Observe real Rust log records through a host-installed capture sink.

use super::{RunnerResult, invalid, text};
use classic_message_core::{ContractEvent, Logger, Message, MessageType};
use serde_json::{Value, json};
use std::sync::{Mutex, Once};

struct Capture;
static CAPTURE: Capture = Capture;
static INSTALL: Once = Once::new();
static RECORDS: Mutex<Vec<Value>> = Mutex::new(Vec::new());

impl log::Log for Capture {
    /// Enable every level so fixture filters cannot silently hide a required call.
    fn enabled(&self, _metadata: &log::Metadata<'_>) -> bool {
        true
    }

    /// Preserve actual level and formatted message while discarding environment-specific targets.
    fn log(&self, record: &log::Record<'_>) {
        RECORDS.lock().expect("capture lock").push(
            json!({"level": record.level().to_string(), "message": record.args().to_string()}),
        );
    }

    /// Records are captured synchronously, so flushing has no deferred work.
    fn flush(&self) {
        // The in-memory sink has no buffered output to flush.
    }
}

/// Execute public logger and formatter operations against input-only fixture values.
pub(super) fn observe(fixture: &Value) -> RunnerResult<Value> {
    INSTALL.call_once(|| {
        log::set_logger(&CAPTURE).expect("install conformance host logger");
        log::set_max_level(log::LevelFilter::Trace);
    });
    RECORDS.lock().expect("capture lock").clear();
    classic_message_core::logging::init();
    classic_message_core::logging::init();
    let logger = Logger::new();
    let mut result = json!({});
    match text(&fixture["operation"])?.as_str() {
        "basic" => {
            let messages = &fixture["messages"];
            logger.info(&text(&messages["info"])?);
            logger.warning(&text(&messages["warning"])?);
            logger.error(&text(&messages["error"])?);
            logger.debug(&text(&messages["debug"])?);
        }
        "extended" => {
            logger.trace(&text(&fixture["trace"])?);
            logger.log(log::Level::Info, &text(&fixture["dynamic"])?);
            logger.log_message(
                &Message::new(text(&fixture["message"])?, MessageType::Warning)
                    .with_title(text(&fixture["title"])?)
                    .with_details(text(&fixture["details"])?),
            );
            result = json!({
                "enabled": [logger.is_info_enabled(), logger.is_debug_enabled(), logger.is_trace_enabled(), logger.is_enabled_for(log::Level::Warn)],
                "name": logger.name(),
                "invalidLog": text(&fixture["invalid"])?.parse::<log::Level>().is_err(),
                "invalidEnabled": text(&fixture["invalid"])?.parse::<log::Level>().is_err()
            });
        }
        "startup" => {
            logger.trace(&text(&fixture["trace"])?);
            let correlation = text(&fixture["correlation"])?;
            logger.log_startup_binding_contract_validated(
                &text(&fixture["contract"])?,
                fixture["checked"]
                    .as_u64()
                    .ok_or_else(|| invalid("checked count"))? as usize,
                Some(&correlation),
            );
            logger.log_startup_binding_contract_failed(
                &text(&fixture["contract"])?,
                &text(&fixture["missing"])?,
                &text(&fixture["failureType"])?,
                &text(&fixture["hint"])?,
                &text(&fixture["error"])?,
                Some(&correlation),
            );
            logger.log_startup_acceleration_status(
                fixture["active"]
                    .as_u64()
                    .ok_or_else(|| invalid("active count"))? as usize,
                fixture["total"]
                    .as_u64()
                    .ok_or_else(|| invalid("total count"))? as usize,
                &text(&fixture["acceleration"])?,
                Some(&correlation),
            );
        }
        "format" => {
            if fixture["severity"] != "warning" {
                return Err(invalid("unsupported contract severity").into());
            }
            let mut event = ContractEvent::new(
                text(&fixture["component"])?,
                text(&fixture["event"])?,
                MessageType::Warning,
                text(&fixture["outcome"])?,
            );
            for (key, value) in fixture["context"]
                .as_object()
                .ok_or_else(|| invalid("context missing"))?
            {
                event = event.with_context(key, text(value)?);
            }
            result["formatted"] = classic_message_core::format_contract_event(&event).into();
        }
        _ => return Err(invalid("unsupported logging operation").into()),
    }
    result["records"] = std::mem::take(&mut *RECORDS.lock().expect("capture lock")).into();
    Ok(result)
}

File: tools/binding_compliance/conformance/families/message_logging.py
"""Exact emitted logging contracts from initialized native adapters."""

import json
from functools import partial

from ..coverage import CoveragePredicate, FamilyCoveragePolicy

_EXPECTED = {
    "basic": {
        "records": [
            {"level": "INFO", "message": "fixture-info"},
            {"level": "WARN", "message": "fixture-warning"},
            {"level": "ERROR", "message": "fixture-error"},
            {"level": "DEBUG", "message": "fixture-debug"},
        ]
    },
    "extended": {
        "records": [
            {"level": "TRACE", "message": "fixture-trace"},
            {"level": "INFO", "message": "fixture-dynamic"},
            {
                "level": "WARN",
                "message": "fixture-title: fixture-message - fixture-details",
            },
        ],
        "enabled": [True, True, True, True],
        "name": "CLASSIC",
        "invalidLog": True,
        "invalidEnabled": True,
    },
    "startup": {
        "records": [
            {"level": "TRACE", "message": "fixture-trace"},
            {
                "level": "INFO",
                "message": "event=classic.startup.binding_contract.validated "
                "severity=info component=integration.startup outcome=success "
                "checked_bindings=2 contract=fixture correlation_id=case",
            },
            {
                "level": "ERROR",
                "message": "event=classic.startup.binding_contract.failed severity=error "
                "component=integration.startup outcome=failure "
                "contract=fixture correlation_id=case error=fixture-error "
                "failure_hint=retry failure_type=fixture "
                "missing_binding=fixture-binding",
            },
            {
                "level": "INFO",
                "message": "event=classic.startup.acceleration.status severity=info "
                "component=integration.startup outcome=success "
                "acceleration_level=fixture active_components=1 "
                "correlation_id=case total_components=2",
            },
        ]
    },
    "format": {
        "records": [],
        "formatted": "event=fixture.event severity=warning component=conformance outcome=ok "
        'count=2 label="two words"',
    },
}
_INPUTS = {
    "basic": {
        "operation": "basic",
        "messages": {
            "info": "fixture-info",
            "warning": "fixture-warning",
            "error": "fixture-error",
            "debug": "fixture-debug",
        },
        "name": "CLASSIC",
    },
    "extended": {
        "operation": "extended",
        "trace": "fixture-trace",
        "dynamic": "fixture-dynamic",
        "message": "fixture-message",
        "title": "fixture-title",
        "details": "fixture-details",
        "invalid": "invalid",
    },
    "startup": {
        "operation": "startup",
        "trace": "fixture-trace",
        "contract": "fixture",
        "checked": 2,
        "correlation": "case",
        "missing": "fixture-binding",
        "failureType": "fixture",
        "hint": "retry",
        "error": "fixture-error",
        "active": 1,
        "total": 2,
        "acceleration": "fixture",
    },
    "format": {
        "operation": "format",
        "component": "conformance",
        "event": "fixture.event",
        "severity": "warning",
        "outcome": "ok",
        "context": {"count": "2", "label": "two words"},
    },
}
_SPECS = {
    "basic": {
        "symbols": ["Logger", "init", "info", "warning", "error", "debug"],
        "ops": [
            None,
            "__init__",
            "createLogger",
            "initLogging",
            "init_logging",
            "info",
            "warning",
            "error",
            "debug",
            "log_info",
            "log_warning",
            "log_error",
            "log_debug",
        ],
    },
    "extended": {
        "symbols": ["Logger"],
        "ops": [
            "trace",
            "log",
            "log_message",
            "name",
            "is_enabled_for",
            "is_info_enabled",
            "is_debug_enabled",
            "is_trace_enabled",
        ],
    },
    "startup": {
        "symbols": [
            "trace",
            "log_startup_binding_contract_validated",
            "log_startup_binding_contract_failed",
            "log_startup_acceleration_status",
        ],
        "ops": [
            "log_trace",
            "log_startup_binding_contract_validated",
            "log_startup_binding_contract_failed",
            "log_startup_acceleration_status",
        ],
    },
    "format": {"symbols": ["format_contract_event"], "ops": ["format_contract_event"]},
}


def _matches(operation, observation):
    """Compare complete log streams and typed results, excluding only environmental log prefixes."""
    return json.dumps(observation, sort_keys=True) == json.dumps(
        _EXPECTED[operation], sort_keys=True
    )


MESSAGE_LOGGING_COVERAGE_POLICY = FamilyCoveragePolicy(
    "message-logging",
    tuple(
        CoveragePredicate(
            id=f"message-logging.{operation}",
            capability_id=f"message-logging.{operation}",
            action=f"message-logging.{operation}",
            observation_family="log-records",
            rust_symbols=tuple(spec["symbols"]),
            runtime_operations=tuple(spec["ops"]),
            matches=partial(_matches, operation),
        )
        for operation, spec in _SPECS.items()
    ),
)


def validate_message_logging_pack(document, root):
    """Reject fixture oracles and require the bounded reviewed logger operations."""
    paths = []
    fixture_root = (root / document["fixtureRoot"]).resolve()
    for scenario in document["scenarios"]:
        reference = scenario["input"].get("fixtureRef")
        operation = scenario["action"].removeprefix("message-logging.")
        if (
            operation not in _INPUTS
            or scenario["input"] != {"fixtureRef": reference}
            or scenario["fixtureRefs"] != [reference]
        ):
            raise ValueError("invalid logging scenario input")
        path = (fixture_root / document["fixtures"][reference]).resolve()
        if not path.is_relative_to(fixture_root):
            raise ValueError("logging fixture escapes root")
        if json.loads(path.read_text(encoding="utf-8")) != _INPUTS[
            operation
        ] or not _matches(operation, scenario["expected"]):
            raise ValueError("invalid logging fixture or expected contract")
        paths.append(path)
    return tuple(paths)
